p2p/transport/webrtc/transport.go

```diff
@@ -0,0 +1,540 @@
+package libp2pwebrtc
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/connmgr"
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/pnet"
+	"github.com/libp2p/go-libp2p/core/sec"
+	tpt "github.com/libp2p/go-libp2p/core/transport"
+	"github.com/libp2p/go-libp2p/p2p/security/noise"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal"
+	"github.com/libp2p/go-libp2p/p2p/transport/webrtc/internal/encoding"
+
+	logging "github.com/ipfs/go-log/v2"
+	ma "github.com/multiformats/go-multiaddr"
+	mafmt "github.com/multiformats/go-multiaddr-fmt"
+	manet "github.com/multiformats/go-multiaddr/net"
+	"github.com/multiformats/go-multihash"
+	pionlogger "github.com/pion/logging"
+
+	"github.com/pion/webrtc/v3"
+)
+
+var log = logging.Logger("webrtc-transport")
+
+var dialMatcher = mafmt.And(mafmt.UDP, mafmt.Base(ma.P_WEBRTC), mafmt.Base(ma.P_CERTHASH))
+
+const (
+	// handshakeChannelNegotiated is used to specify that the
+	// handshake data channel does not need negotiation via DCEP.
+	// A constant is used since the `DataChannelInit` struct takes
+	// references instead of values.
+	handshakeChannelNegotiated = true
+	// handshakeChannelID is the agreed ID for the handshake data
+	// channel. A constant is used since the `DataChannelInit` struct takes
+	// references instead of values. We specify the type here as this
+	// value is only ever copied and passed by reference
+	handshakeChannelID = uint16(0)
+)
+
+// timeout values for the peerconnection
+// https://github.com/pion/webrtc/blob/v3.1.50/settingengine.go#L102-L109
+const (
+	DefaultDisconnectedTimeout = 5 * time.Second
+	DefaultFailedTimeout       = 25 * time.Second
+	DefaultKeepaliveTimeout    = 2 * time.Second
+)
+
+type WebRTCTransport struct {
+	webrtcConfig webrtc.Configuration
+	rcmgr        network.ResourceManager
+	privKey      ic.PrivKey
+	noiseTpt     *noise.Transport
+	localPeerId  peer.ID
+
+	// timeouts
+	peerConnectionTimeouts ICETimeouts
+
+	// in-flight connections
+	maxInFlightConnections uint32
+}
+
+var _ tpt.Transport = &WebRTCTransport{}
+
+type Option func(*WebRTCTransport) error
+
+type ICETimeouts struct {
+	Disconnect time.Duration
+	Failed     time.Duration
+	Keepalive  time.Duration
+}
+
+// WithPeerConnectionIceTimeouts sets the ice disconnect, failure and keepalive timeouts
+func WithPeerConnectionIceTimeouts(timeouts ICETimeouts) Option {
```

Comment:
Sorry for asking this again, but I don't understand what we're doing here. Why do we need to set _any_ ICE parameters. WebRTC browser-to-server doesn't use STUN nor TURN, so I don't understand what an ICE disconnect or keepalive timeout even means.

Is this somehow configuring keepalives for the resulting SCTP connection? If so, what's the difference between "failed" and "disconnected"?

API-wise, I think these should be 2 (or 3, if there's an important different between "failed" and "disconnected") separate options, not a struct.

Reply:
The ICE parameters allow us to tune when to close a connection when we have not received any responses from the remote peer. This could be the remote not sending any data or not responding to any STUN keepalives that are sent.

The timeouts are described [here](https://pkg.go.dev/github.com/pion/webrtc/v3#SettingEngine.SetICETimeouts). If a peer does not respond to timeouts until the `disconnectedTimeout`, the connection is marked a disconnected. However, if the peer responds before `failureTimeout`, the connection is again marked as connected. In case the failure timeout is also breached, the connection is marked as failed an should be closed.

Reply:
libp2p doesn't have the distinction between failed and disconnected, and I see no reason to expose an API for that. This package should do the right thing, so that users of this transport don't have to worry about it.

For the keep-alive, I suggest we use the same value that QUIC uses (15s). This doesn't need to be configurable. The idle timeout (which, iiuc, would correspond to the `failureTimeout`) should be 30s, also doesn't need to be configurable. Not sure what that means for the `disconnectedTimeout`, any suggestions?